dts/bindings/ethernet/xlnx,gem.yaml

```diff
@@ -0,0 +1,353 @@
+#
+# Copyright (c) 2021, Weidmueller Interface GmbH & Co. KG
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: Xilinx GEM Ethernet controller
+
+compatible: "xlnx,gem"
+
+include: ethernet.yaml
+
+properties:
+    reg:
+      required: true
+
+    interrupts:
+      required: true
+
+    clock-frequency:
+      type: int
+      required: true
+      description: TX Clock base frequency
+        Specifies the base clock frequency from which the GEM's TX clock
+        frequency will be derived using two dividers in the respective GEM's
+        clock control register in the CRL_APB. The GEM's TX clock frequency
+        is determined by the current link speed reported by the PHY, to
+        which it will be adjusted at run-time. Therefore, the value of this
+        item must be set to the clock frequency of the PLL supplying the
+        respective GEM's TX clock - by default, this is the IO PLL.
+
+    clk-ctrl-reg:
```

Comment:
Can this be changed to use a clock driver instead of using a hardcoded address to another block?

Reply:
I agree, it looks like this should probably take a phandle to another node with compatible fixed-clock. cc @galak

Reply:
This whole clock-thing is a bit of a complicated issue.

In the Xilinx SoCs, all of the clock sources, frequencies, multipliers/dividers etc. relating to the Processor System, integrated peripherals and the programmable logic part have to be configured as part of the FPGA synthesis project in the Vivado IDE (which unfortunately results in redundant work, the whole clock config has to be specified first in Vivado and then in the device tree of the target board, with maybe some defaults at the SoC level of the DT matching the most common configurations). A by-product of the generated FPGA bitstream is an auto-generated C source file which contains loads of register read&write commands which effectively replicate those settings into what eventually becomes the first stage bootloader binary, built by the Xilinx SDK. Therefore, driver for Xilinx peripherals shouldn't mess with any clock configurations during start up, as the system is always pre-configured when booting into the OS.

The only exception to that rule are input clock prescalers that can be found for example in the UART for the generation of the configured baud rate, CAN has such a prescaler as well, and in the Ethernet Controller, a prescaler of the input clock is required for the TX clock generation based on the link speed reported by the PHY. The peripheral drivers need not know or actively alter the setting which PLL drives their respective device, they just need to know the actual input frequency, regardless of the clock source, based on which the respective prescaler is calculated, and this value has to be written to the respective peripheral's clock configuration register without modifying any other contents of that register. Therefore, the additional "clock-frequency" item in the DT binding is sufficient for the calculation part and is consistent with how input clocks are specified for other device classes such as UARTs, timers etc. As long as there's no "reg" property contained in the "fixed-clock" class (see explanation below), I don't think that there's any real benefit in referencing a separate clock node in the DT for the specification of the plain source clock frequency.

Regarding the clock configuration register to which the prescaler eventually has to be written, there's a bit of an unfortunate design decision by Xilinx: the peripheral clock configuration register are one (Zynq, SLCR) or three (UltraScale, CRF_APB, CRL_APB, IOU_SLCR) separate register spaces, so that each peripheral's clock configuration register is not contained within the register space specified by each peripheral's "reg" property. My initial idea on how to solve this issue was to create one/three clock configuration node(s) in the DT, based on some existing nodes of such type I think I found in the STM32 device trees. This would have allowed to specify (most likely at the board level) the HW oscillator frequency and all the routing / multiplier / divider / clocking scheme informations that configure the entire system's clocks. I added the calculation of frequencies derived from those base settings in the SoC init code, and there was getters for everything at that level, too. Setters were provided for each peripheral's prescaler where required, without the driver having to know its associated register address (this was resolved via the peripheral's base address). The only redundancy (aside from having to set up everything in Vivado first, then having to copy all the data from the synthesis report to the board's DT) then would have been SYS_CLOCK_HW_CYCLES_PER_SEC. The GEM driver in development, and also future implementations such as a CAN, I2C, ... driver, could have just called a getter to obtain the respective input clock frequency, could have calculated the prescaler, and could have called a setter with the device's base address and the result of the calculation as parameter. That way, no knowledge of bit masking and shifting in the config register (differs slightly between Zynq and UltraScale here and there) would have been required either, one #ifdef less. The moment when all this fell apart was when I realized that even if I shifted the existing UART driver over to this scheme, UART devices always require a "clock-frequency" value. At that point, there would have been an inconsistency, so I threw it all out again and opted to make it all consistent again by having the GEM have a "clock-frequency" property as well.

This still doesn't solve the control register issue, though. The register address could be determined within the driver, but that would require a nested switch first over the current SoC family, then over the current device's base address. Admittedly, that's not a great solution - as the driver currently is, the SoC family switch is already back in for writing to the clock control register due to the slightly different bit positions. So if a "fixed-clock" node can have a "reg" property without having to introduce an additional (if optional) property that's useless pretty much every other use case, then I'd agree that referencing a separate node that contains both the frequency and the register location would be the best way to solve the issue. But if that is not the case, I'm short on ideas on how to solve the issue elegantly.

Reply:
fixed-clock indeed has no reg; you can see this here:

https://docs.zephyrproject.org/latest/reference/devicetree/bindings/clock/fixed-clock.html#dtbinding-fixed-clock

I'd rather not block this PR on that detail then. It seems like this is a pretty special case, and your users are going to have to jump through enough hoops as it is.

However, while I was looking at the bindings index, I realized that the binding added for this compatible doesn't follow the usual practice for multiline descriptions, which is to use:

```yaml
  description: |
    My very long description
    goes here
```

This is a bit of a game of whack-a-mole to keep clean, but I've just done another pass while I was here: https://github.com/zephyrproject-rtos/zephyr/pull/36189

Would you mind following suit?

Reply:
I'll change the multiline formatting, no problem!